### DISCUSSION TITLE
Getting Preview to work without Next.js opting-out to SSR

### DISCUSSION THREAD

**Discussion by riccardolardi** [craftcms/cms#14903] [Q&A] [answered]
Hi all. Is it possible to make Craft preview work without passing back the token? I'm using Nextjs and with the current App Router it's not possible to read the received token **without opting out to dynamic rendering (SSR) because the query string cannot be known ahead of time**. See here for reference https://nextjs.org/docs/app/api-reference/file-conventions/page#searchparams-optional

If it were possible to e.g. define a "secret" instead, that stays static and can be set as an environment variable for example, that would solve the problem I believe.

I think this is a very common use case for people using Next.js App Router.

I'm not sure if CraftCMS 5 has changed in this regard and offers new ways of solving this?

**Comment by brandonkelly** [comment, marked as answer]
Each token is unique to the draft being previewed. At any given time there could be multiple drafts of each entry, with one or more people accessing each draft. How would Craft know which draft(s) to load in place of the live entries, if all it has to go on is a single shared secret between itself and the Next.js app?

**Comment by riccardolardi** [comment]
Thanks a lot @brandonkelly I understand. I'll try to pass the token in a cookie from the API route in Next.js instead of using URL query parameters.

**Reply by louderthan10** [reply]
> Thanks a lot @brandonkelly I understand. I'll try to pass the token in a cookie from the API route in Next.js instead of using URL query parameters.

I put together our [Next.js starter](https://github.com/craftcms/starter-next), and have to say it was much more difficult to get Live Preview going in Next than in Nuxt or Astro for this reason. Using a cookie to pass the token is an interesting approach that I didn't consider. Going to revisit that approach myself and see if I can simplify things further.

**Reply by stuartmitchelhill** [reply]
@louderthan10 i've implemented the cookie solution it works quite nicely and feels cleaner then having to call searchParams in every template, however still doesn't solve the issue of SSR as this still forces dynamic rendering being a dynamic API.

Would be great if you had any insight on preserving SSR while still being able to utilise live preview? I'm still testing a few ideas and will let you know if I find anything that works well

**Comment by stuartmitchelhill** [comment]
@brandonkelly is a potential workaround for this to allow a static preview token to be set in the craft config, and if so, disable multiple drafts and lock editing when another user is already accessing the entry?

This would be really handy to keep ability to pre-render in Next.js apps but understand that it would require a significant amount of work to implement depending on how drafts and previews are currently handled, and that your product is independent of Next
